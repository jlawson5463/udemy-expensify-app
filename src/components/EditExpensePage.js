import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startRemoveExpense, editExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense({id: this.props.expense.id, updates: expense}); 
        this.props.history.push('/');
    }
    onRemove = (id) => {
        this.props.startRemoveExpense(id);
        this.props.history.push('/');
    }
    render(){
        return (
            <div>
            <ExpenseForm
                expense={this.props.expense}
                onSubmit={this.onSubmit} 
            />
            <button onClick={(e) => {
                this.onRemove({id: this.props.expense.id});
            }}>Remove
            </button>
        </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        editExpense: (expense) => dispatch(editExpense(expense)),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);