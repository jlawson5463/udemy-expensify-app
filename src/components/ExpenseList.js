import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {props.expenses.map((e, i) => 
            <ExpenseListItem {...e} key={i}/>
        )}
    </div>
);

// below is where we set up what part of the store we want this component to be able to access.
// as the provider passes the entire store to every component we need to narrow down what it can see
// the expenses in the below object can be called anything.
// just a placeholder for prop names to store things from the state
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

// passing state into the redux store using the connect api. 
// it returns what we set it to - so expenses in this case.
// passes them as props to what ever we pass in to the 2nd place.
// now can display them by accessing the props.
export default connect(mapStateToProps)(ExpenseList);
