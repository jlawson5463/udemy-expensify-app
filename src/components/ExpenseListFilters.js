import React from 'react';
import { connect } from 'react-redux';
import { setFilterText, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import { css, withStyles, withStylesPropTypes } from 'react-with-styles';

export class ExpenseListFilters extends React.Component {
    state = {
        calenderFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange = (calenderFocused) => {
        this.setState(() => ({ calenderFocused }))
    }
    onSortChange = (e) => {
        e.target.value == "date" ? this.props.sortByDate() : this.props.sortByAmount()
    }
    onTextChange = (e) => {
        this.props.setFilterText(e.target.value);
    }
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input  type='text' 
                                className="text-input"
                                placeholder="Search expenses"
                                value={this.props.filters.text}
                                onChange={this.onTextChange} />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            onChange={this.onSortChange}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDateId="startDateId"
                            startDate={this.props.filters.startDate}
                            endDateId="endDateId"
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calenderFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            isOutsideRange={(day) => false}
                            numberOfMonths={1}
                            displayFormat={('DD MMM YY')}
                        >
                        </DateRangePicker>
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => {
    return {
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setFilterText: (value) => dispatch(setFilterText(value))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);