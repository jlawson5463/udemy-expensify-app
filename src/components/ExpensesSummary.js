import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expenseTotal }) => {
   const formattedTotal = numeral(expenseTotal / 100).format('£0,0.00');
    return (
        <div>
            <h1>Viewing {expenseCount} {expenseCount === 1 ? 'expense ' : 'expenses '} 
                totalling £{formattedTotal}</h1>
        </div>
    )
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expenseCount: visibleExpenses.length,       
        expenseTotal: expensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps, undefined)(ExpensesSummary);
