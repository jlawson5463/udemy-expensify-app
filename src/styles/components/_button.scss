.button {
background: $blue;
border: none;
border-radius: 3px;
display: inline-block;
font-size: $font-size-large;
font-weight: 300;
line-height: 1;
padding: $s-size;
color: white;
text-decoration: none;
}

.button--link {
    background: none;
}

.button--edit {
    background: $dark-grey;
}