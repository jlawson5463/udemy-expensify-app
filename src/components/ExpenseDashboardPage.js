import React from 'react';
import ConnectedExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = (props) => {
    console.log(props)
    return (
        <div>
            <ExpenseListFilters />
            <ConnectedExpenseList />
        </div>
        
    );
}

export default ExpenseDashboardPage;