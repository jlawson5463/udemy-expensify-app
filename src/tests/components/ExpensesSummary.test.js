import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render expense summary for 1 expense', () => {
    const expenseTotal = 100;
    const expenseCount = 1;
    const wrapper = shallow(<ExpensesSummary expenseCount={expenseCount} expenseTotal={expenseTotal}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render expense summary for multiple expenses', () => {
    const expenseTotal = 2000;
    const expenseCount = 3;
    const wrapper = shallow(<ExpensesSummary expenseCount={expenseCount} expenseTotal={expenseTotal}/>);
    expect(wrapper).toMatchSnapshot();
});
