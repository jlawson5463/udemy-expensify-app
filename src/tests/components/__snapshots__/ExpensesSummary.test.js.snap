// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary for 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense 
    totalling £
    1.00
  </h1>
</div>
`;

exports[`should render expense summary for multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses 
    totalling £
    20.00
  </h1>
</div>
`;
